common/kotlinx-coroutines-core-common/src/JobSupport.kt

```diff
@@ -568,155 +604,220 @@ internal open class JobSupport constructor(active: Boolean) : Job, SelectClause0
-            when {
-                state !is Finishing && hasOnFinishingHandler(proposedUpdate) -> null // unless it has onFinishing handler
-                tryFinalizeState(state, proposedUpdate, mode) -> return COMPLETING_COMPLETED
-                else -> return@loopOnState
-            }
-            val list = state.list ?: // must promote to list to correctly operate on child lists
-            when (state) {
-                is Empty -> {
-                    promoteEmptyToNodeList(state)
-                    return@loopOnState // retry
-                }
-                is JobNode<*> -> {
-                    promoteSingleToNodeList(state)
-                    return@loopOnState // retry
-                }
-                else -> error("Unexpected state with an empty list: $state")
-            }
-            // cancel all children in list on exceptional completion
-            if (proposedUpdate is CompletedExceptionally)
-                child?.cancelChildrenInternal(proposedUpdate.cause)
-            // switch to completing state
-            val cancelled = (state as? Finishing)?.cancelled ?: (proposedUpdate as? Cancelled)
-            val completing = Finishing(list, cancelled, true)
-            if (_state.compareAndSet(state, completing)) {
-                (state as? Finishing)?.transferExceptions(completing)
-                if (state !is Finishing) onFinishingInternal(proposedUpdate)
-                if (child != null && tryWaitForChild(child, proposedUpdate))
-                    return COMPLETING_WAITING_CHILDREN
-                if (tryFinalizeState(completing, proposedUpdate, mode))
-                    return COMPLETING_COMPLETED
-            }
+            COMPLETING_COMPLETED -> return true
+            COMPLETING_WAITING_CHILDREN -> return false
+            COMPLETING_RETRY -> return@loopOnState
+            else -> error("unexpected result")
         }
     }
 
-    private tailrec fun ChildJob.cancelChildrenInternal(cause: Throwable) {
-        childJob.cancel(JobCancellationException("Child job was cancelled because of parent failure", cause, childJob))
-        nextChild()?.cancelChildrenInternal(cause)
+    private fun tryMakeCompleting(state: Any?, proposedUpdate: Any?, mode: Int): Int {
+        if (state !is Incomplete)
+            return COMPLETING_ALREADY_COMPLETING
+        // FAST PATH -- no children to wait for && simple state (no list) && not failing => can complete immediately
+        // Failures always have to go through Finishing state to serialize exception handling
+        if ((state is Empty || state is JobNode<*>) && state !is ChildJob && proposedUpdate !is CompletedExceptionally) {
+            if (!tryFinalizeSimpleState(state, proposedUpdate, mode)) return COMPLETING_RETRY
+            return COMPLETING_COMPLETED
+        }
+        // get state's list or else promote to list to correctly operate on child lists
+        val list = tryPromoteToList(state) ?: return COMPLETING_RETRY
```

Comment:
I've optimization Empty -> Failing state transition, but we cannot optimize single JobNode -> Failing transtiion in this way. A single JobNode must be promoted to NodeList, because this transition works correctly only when it is pointed to by state.